### PR TITLE
Gemma: enabled HPU Graphs and Flash Attention 

### PR DESCRIPTION
**Pull request by dsmertin**
# What does this PR do?

This PR fixes HPU Graphs usage and Flash Attention for Gemma model.
Changes are based on Starcoder 2 and Qwen 2 implementations.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yafshar** [comment]
@dsmertin can you re-base this PR with the main?

**Review by imangohari1 — CHANGES_REQUESTED** [review]
Hi @dsmertin 
Thank you for this PR. 
Could you please address the followings and push the changes?
- Pls. rebase/sync on top of OH main. Currently 72 commits behind. 
- Pls. make sure to run `make style` 
- Pls. share the results of these gemma before/after these changes. 
- We need to test/update CI tests with these changes.

File: optimum/habana/transformers/models/gemma/modeling_gemma.py [hunk 1]
```diff
@@ -40,91 +43,382 @@
...
+    
+    def gaudi_flash_attn_v1(self, query_layer, key_layer, value_layer, attention_mask, dropout_rate, q_block_size):
+        """
+        Gaudi version of Flash Attention V1 to support long sequence at prompt phase
+        Causal mask is not supported in this optimization
+        """
+        q_len = query_layer.size(-2)
+        q_tiles = (q_len // q_block_size) if (q_len % q_block_size == 0) else math.ceil(q_len / q_block_size)
+        q_padding = q_tiles * q_block_size - q_len
+        query_layer = F.pad(query_layer, (0, 0, 0, q_padding), "constant", 0)
```
Comment (imangohari1) [inline optimum/habana/transformers/models/gemma/modeling_gemma.py:178]: make style here complains the F is not defined. 
Pls. take a look.

File: optimum/habana/transformers/models/gemma/modeling_gemma.py [hunk 2]
```diff
@@ -40,91 +43,382 @@
...
+        if attention_mask is not None:
+            attention_mask = F.pad(attention_mask, (0, 0, 0, q_padding), "constant", -10000.0)
```
Comment (imangohari1) [inline optimum/habana/transformers/models/gemma/modeling_gemma.py:180]: same here about `make style`

File: optimum/habana/transformers/models/gemma/modeling_gemma.py [hunk 3]
```diff
@@ -40,91 +43,382 @@
...
+        The only differences are:
+        - add new args token_idx
+        - optimize KV cache
+        - add new args attn_softmax_bf16
+        - add new args reuse_cache
+        - add new args use_flash_attention
+        - add new arg flash_attention_recompute
+        """
+        if "padding_mask" in kwargs:
+            warnings.warn(
```
Comment (imangohari1) [inline optimum/habana/transformers/models/gemma/modeling_gemma.py:224]: make style here complains the `warnings` are not defined. 
shouldn't this be a `logger.warning_once`

**Review by imangohari1 — CHANGES_REQUESTED** [review]
I did a bit more testing here and added few more comments. 
this PR also has merge conflict with main. 
Please make sure to address them during the rebase, and `make style`. 
Thanks.

**Review by imangohari1 — CHANGES_REQUESTED** [review]
I fixed some issues and updated the code with `make style` and added `--use_flash_attention` to gemma CI test. 
Please apply the attach fix with `git am < 0001*` and push. 
Thanks. 
[0001-fix-gemma-make-style.-minor-fixes.patch](https://github.com/user-attachments/files/16971015/0001-fix-gemma-make-style.-minor-fixes.patch)

**Comment by yafshar** [comment]
@dsmertin can you address the comments?

**Comment by imangohari1** [comment]
> @yafshar @imangohari1 This PR was created a month and a half ago for different version of the software stack. There was a difference and improvement in performance for this model, but I'm not sure if it's still the case. My colleague told me that there is not much difference for 17.1 version and I need to double check it now. If there'll be no improvement, I'll close it then. Otherwise, I'll go through the comments.

@dsmertin 
Thanks. 
I have tested this PR with the latest release and we would like this PR to be included for the next release. 

Please work through the changes I have suggested here and push them so we can test this further on the RC. 

thanks. 

> I fixed some issues and updated the code with `make style` and added `--use_flash_attention` to gemma CI test. Please apply the attach fix with `git am < 0001*` and push. Thanks. [0001-fix-gemma-make-style.-minor-fixes.patch](https://github.com/user-attachments/files/16971015/0001-fix-gemma-make-style.-minor-fixes.patch)

**Review by imangohari1 — APPROVED** [review]
LGTM!. 
@regisss please take a look. Thanks

**Comment by regisss** [comment]
Ci failed with:
```
E     File "/root/workspace/optimum/habana/transformers/models/gemma/modeling_gemma.py", line 552
E       use_cache = use_cache if use_cache is not None else self.config.use_cache
E   IndentationError: unexpected indent
```

File: optimum/habana/transformers/trainer.py [hunk 4]
```diff
@@ -971,9 +971,9 @@ def hpu_deepspeed_checkpointing(function, *checkpoint_args, use_reentrant: Optio
                 if step % args.gradient_accumulation_steps == 0:
                     self.control = self.callback_handler.on_step_begin(args, self.state, self.control)
 
-                # attn_softmax_bf16 and use_flash_attention is enabled only for llama, qwen2 and starcoder2
+                # attn_softmax_bf16 and use_flash_attention is enabled only for llama, qwen2 and gemma
```
Comment (regisss) [inline optimum/habana/transformers/trainer.py:974]:
```suggestion
                # attn_softmax_bf16 and use_flash_attention is enabled only for llama, qwen2, starcoder2 and gemma
```

File: optimum/habana/transformers/trainer.py [hunk 5]
```diff
@@ -1878,9 +1878,9 @@ def evaluation_loop(
                 if batch_size is None:
                     batch_size = observed_batch_size
 
-            # attn_softmax_bf16 and use_flash_attention are enabled only for llama, qwen2 and starcoder2
+            # attn_softmax_bf16 and use_flash_attention are enabled only for llama, qwen2 and gemma
```
Comment (regisss) [inline optimum/habana/transformers/trainer.py:1881]:
```suggestion
            # attn_softmax_bf16 and use_flash_attention are enabled only for llama, qwen2, starcoder2 and gemma
```

File: optimum/habana/transformers/generation/utils.py [hunk 6]
```diff
@@ -1127,7 +1132,7 @@ def generate(
                 model_kwargs["kv_cache_len"] = calculated_max_length
                 model_kwargs["kv_cache_pad_len"] = generation_config.max_new_tokens
 
-            if self.config.model_type in ["llama", "falcon", "mistral", "qwen2", "gptj", "starcoder2"]:
+            if self.config.model_type in ["llama", "falcon", "mistral", "qwen2", "starcoder2", "gemma"]:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1135]: Why removing GPT-J from this list?

**Pull request merged**